lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -1984,47 +1984,30 @@ def plot_trisurf(self, *args, **kwargs):
             args = args[1:]
 
         triangles = tri.get_masked_triangles()
-        xt = tri.x[triangles][..., np.newaxis]
-        yt = tri.y[triangles][..., np.newaxis]
-        zt = z[triangles][..., np.newaxis]
+        xt = tri.x[triangles]
+        yt = tri.y[triangles]
+        zt = z[triangles]
 
-        verts = np.concatenate((xt, yt, zt), axis=2)
-
-        # Only need these vectors to shade if there is no cmap
-        if cmap is None and shade:
-            totpts = len(verts)
-            v1 = np.empty((totpts, 3))
-            v2 = np.empty((totpts, 3))
-            # This indexes the vertex points
-            which_pt = 0
-
-        colset = []
-        for i in xrange(len(verts)):
-            avgzsum = verts[i,0,2] + verts[i,1,2] + verts[i,2,2]
-            colset.append(avgzsum / 3.0)
-
-            # Only need vectors to shade if no cmap
-            if cmap is None and shade:
-                v1[which_pt] = np.array(verts[i,0]) - np.array(verts[i,1])
-                v2[which_pt] = np.array(verts[i,1]) - np.array(verts[i,2])
-                which_pt += 1
-
-        if cmap is None and shade:
-            normals = np.cross(v1, v2)
-        else:
-            normals = []
+        # verts = np.stack((xt, yt, zt), axis=-1)
```

Comment:
Are they 1D or 2D? `np.vstack` or `np.dstack` could also be used in those cases.

Reply:
Yep, this was a new change since you last reviewed. Both vstack and dstack have undesirable semantics of guessing the array sizes, and using `concatenate` is more precise.

This isn't the only place that `np.stack` is mentioned in a comment above a `np.concatenate((a[..., None], ...))`, so might be worth backporting to the very simple
```
def stack(arrays, axis):
    return np.concatenate([arr[..., np.newaxis] for arr in arrays], axis)
```

Not something I want to involve in this PR though

Reply:
You can just stick the definition in https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/cbook/_backports.py if you change your mind.

Reply:
Oh, and to answer your question, 2D - `(N, edges_in_triangle)`

Reply:
We do have a `cbook._backports`; not sure how strong the need is for it without having done any looking.

Reply:
I reckon there're a substantial number of cases - I can put together a PR once this is merged

Reply:
Looking again, a `stack_last` function is all that ever seems to be needed, implemented as the above - so that wouldn't belong in `_backports`. Suggestions of where to put such a helper?

Reply:
cbook would be the place.
But given that we'd probably just be using stack(..., -1) "if it was available", I'd just backport stack.
Let us know if you want to do this in this PR or a separate one.

Reply:
Definitely a separate one